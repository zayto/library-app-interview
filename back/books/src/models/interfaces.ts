import { Types } from "mongoose";

export enum BookStatusEnum {
  AVAILABLE = 'AVAILABLE',
  BORROWED = 'BORROWED',
  IN_TRANSIT = 'IN_TRANSIT',
  TO_BE_PICKED = 'TO_BE_PICKED',
  DAMAGED = 'DAMAGED',
  LOST = 'LOST'
}

export enum BookRefStatusEnum {
  AVAILABLE = 'AVAILABLE',
  TO_ORDER_BACK = 'TO_ORDER_BACK',
  UNAVAILABLE = 'UNAVAILABLE',
}

export interface IBookRequest {
  reference: string;
  status: BookStatusEnum;
}

export interface IBookResponse {
  _id: Types.ObjectId;
  reference: Types.ObjectId; // A reference to the actual book (bookRefs collection)
  status: BookStatusEnum;
  owner: Types.ObjectId; // A reference to the User who currently owns the book (users collection)
}

export interface IBookRefRequest {
  title: string;
  author: string;
  excerpt: string;
  status: BookRefStatusEnum;
  quantity: number;
}

export interface IBookRefResponse {
  _id: Types.ObjectId;
  title: string;
  author: string;
  excerpt: string;
  status: BookRefStatusEnum;
  available: number; // Quantity available in the shelves
  totalQuantity: number; // Total quantity owned by the library for this reference
}
