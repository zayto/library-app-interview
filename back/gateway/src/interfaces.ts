// TODO Find a way to retrieve the IUserResponse from Users module directly to avoid duplication
export interface IUserResponse {
  id: number;
  firstName: string;
  lastName: string;
  books: string[];
}
