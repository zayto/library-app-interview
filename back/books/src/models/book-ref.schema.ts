import { Types, Document } from 'mongoose';
import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { BookRefStatusEnum } from './interfaces';

export type BookDocument = BookRef & Document;

@Schema()
export class BookRef {
  _id: Types.ObjectId;

  @Prop({ required: true })
  author: string;

  @Prop({ required: true })
  title: string;

  @Prop({ default: '' })
  excerpt: string;

  @Prop({ required: true })
  status: BookRefStatusEnum;

  @Prop({ required: true })
  available: number;

  @Prop({ required: true })
  totalQuantity: number;
}

export const BookRefSchema = SchemaFactory.createForClass(BookRef);
