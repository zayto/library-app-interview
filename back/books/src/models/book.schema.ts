import { SchemaTypes, Types, Document } from 'mongoose';
import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { BookStatusEnum } from './interfaces';

export type BookDocument = Book & Document;

@Schema()
export class Book {
  @Prop({ required: true })
  status: BookStatusEnum;

  @Prop({ type: SchemaTypes.ObjectId, ref: 'Reference' })
  reference: Types.ObjectId;

  @Prop({ type: SchemaTypes.ObjectId, ref: 'User' })
  owner: Types.ObjectId;
}

export const BookSchema = SchemaFactory.createForClass(Book);
