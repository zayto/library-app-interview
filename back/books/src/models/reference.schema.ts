import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { Document } from 'mongoose';
import { BookRefStatusEnum } from './interfaces';

export type ReferenceDocument = Reference & Document;

@Schema()
export class Reference {
  @Prop({ required: true })
  title: string;
  
  @Prop({ required: true })
  author: string;

  @Prop({ required: true })
  excerpt: string;

  @Prop()
  status: BookRefStatusEnum;

  @Prop({ required: true })
  available: number;
  
  @Prop({ required: true })
  totalQuantity: number;
}

export const ReferenceSchema = SchemaFactory.createForClass(Reference);
