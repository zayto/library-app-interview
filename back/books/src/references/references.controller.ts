import { Controller } from '@nestjs/common';

@Controller('references')
export class ReferencesController {}
