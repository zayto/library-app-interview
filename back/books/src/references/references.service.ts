import { Injectable, Logger } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model, Types } from 'mongoose';
import {
  BookRefStatusEnum,
  IBookRefRequest,
  IBookRefResponse,
} from '../models/interfaces';
import { Reference, ReferenceDocument } from '../models/reference.schema';

@Injectable()
export class ReferencesService {
  constructor(
    @InjectModel(Reference.name) private refModel: Model<ReferenceDocument>,
  ) {}

  public async getBookReferenceById(
    refId: string,
  ): Promise<IBookRefResponse | null> {
    Logger.log(`GetBookReferenceById called with id=${refId}`);
    return await this.refModel.findById<IBookRefResponse>(refId);
  }

  public async getAllBooksRefs(): Promise<IBookRefResponse[]> {
    Logger.log('GetAllBookRefs called');
    return this.refModel.find().exec();
  }

  public async createBookRef(
    bookRef: IBookRefRequest,
  ): Promise<IBookRefResponse> {
    Logger.log(`CreateBookRef called with payload`, {
      bookRef: JSON.stringify(bookRef),
    });

    const createdBookRef = await this.refModel.create({
      ...bookRef,
      available: 0,
      totalQuantity: 0,
      status: BookRefStatusEnum.TO_ORDER_BACK, // Represents a new entry but with 0 quantity, we should order some books
    });
    return createdBookRef;
  }

  public async borrowBookAndUpdateBookRef(
    bookRef: IBookRefResponse,
  ): Promise<void> {
    const { _id: bookRefId } = bookRef;
    if (!this.isBookRefAvailable(bookRef)) {
      throw new Error(
        `Book with ref=${bookRefId} is not available to be borrowed`,
      );
    }

    const refToUpdate: Partial<IBookRefResponse> = {
      title: bookRef.title,
      author: bookRef.author,
      excerpt: bookRef.excerpt,
      status:
        bookRef.available > 0
          ? BookRefStatusEnum.AVAILABLE
          : BookRefStatusEnum.UNAVAILABLE,
      available: bookRef.available - 1,
    };
    // Update availability status and quantity after borrowing
    await this.refModel.findOneAndUpdate(bookRefId, refToUpdate);
  }

  public async addNewBookQuantity(
    bookRefId: string,
    quantity: number,
  ): Promise<IBookRefResponse> {
    if (quantity <= 0) {
      throw new Error(
        `Invalid quantity provided (quantity=${quantity}) to add for bookRef with id=${bookRefId}`,
      );
    }
    const bookRef = await this.refModel.findById(bookRefId, null);
    bookRef.available += quantity;
    bookRef.totalQuantity += quantity;
    if (bookRef.available > 0) {
      bookRef.status = BookRefStatusEnum.AVAILABLE;
    }
    const updated = await bookRef.save();
    Logger.log(
      `Added quantity ${quantity} to ref with id=${bookRefId}, new availableQuantity=${updated.available}, new totalQuantity=${updated.totalQuantity}`,
    );

    return updated;
  }

  private isBookRefAvailable(bookRef: IBookRefResponse): boolean {
    return (
      bookRef.available > 0 && BookRefStatusEnum.AVAILABLE === bookRef.status
    );
  }
}
