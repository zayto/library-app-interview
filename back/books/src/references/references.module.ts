import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { ReferencesController } from './references.controller';
import { ReferencesService } from './references.service';
import { BookRef, BookRefSchema } from '../models/book-ref.schema';
import { ConfigModule } from '@nestjs/config';

@Module({
  imports: [
    ConfigModule.forRoot(),
    MongooseModule.forRoot(
      `mongodb://user:password@${process.env.MONGO_HOST}/library`,
    ),
    MongooseModule.forFeature([{ name: BookRef.name, schema: BookRefSchema }]),
  ],
  controllers: [ReferencesController],
  providers: [ReferencesService],
  exports: [MongooseModule],
})
export class ReferencesModule {}
