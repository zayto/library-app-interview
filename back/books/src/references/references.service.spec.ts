import { getModelToken } from '@nestjs/mongoose';
import { Test, TestingModule } from '@nestjs/testing';
import { Model, Types } from 'mongoose';
import { BookRefStatusEnum, IBookRefResponse } from '../models/interfaces';
import { Reference } from '../models/reference.schema';
import { ReferencesService } from './references.service';

describe('ReferencesService', () => {
  let refService: ReferencesService;
  let model: Model<Reference>;

  const mockRef: IBookRefResponse = {
    _id: new Types.ObjectId('6209880b6f43a7e034c84943'),
    author: 'Author',
    title: 'Title',
    excerpt: 'Excerpt',
    available: 1,
    totalQuantity: 1,
    status: BookRefStatusEnum.AVAILABLE,
  };

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        ReferencesService,
        {
          provide: getModelToken(Reference.name),
          useValue: {
            new: jest.fn().mockResolvedValue(mockRef),
            constructor: jest.fn().mockResolvedValue(mockRef),
            find: jest.fn().mockReturnValue({ exec: () => [mockRef] }),
            create: jest.fn(),
            exec: jest.fn(),
          },
        },
      ],
    }).compile();

    model = module.get<Model<Reference>>(getModelToken('Reference'));
    refService = module.get<ReferencesService>(ReferencesService);
  });

  it('should be defined', () => {
    expect(refService).toBeDefined();
  });

  it('should return all bookRefs', async () => {
    const refs = await refService.getAllBooksRefs();
    expect(refs).toMatchSnapshot();
  });
});
