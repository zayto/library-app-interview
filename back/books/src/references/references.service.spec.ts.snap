// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReferencesService should return all bookRefs 1`] = `
Array [
  Object {
    "_id": "6209880b6f43a7e034c84943",
    "author": "Author",
    "available": 1,
    "excerpt": "Excerpt",
    "status": "AVAILABLE",
    "title": "Title",
    "totalQuantity": 1,
  },
]
`;
